src/Compilers/CSharp/Portable/Declarations/MergedNamespaceDeclaration.cs

```diff
@@ -139,43 +140,102 @@ private ImmutableArray<MergedNamespaceOrTypeDeclaration> MakeChildren()
 
             var children = ArrayBuilder<MergedNamespaceOrTypeDeclaration>.GetInstance();
 
-            if (namespaces != null)
+            addNamespacesToChildren(namespaces, allNamespacesHaveSameName, children);
+            addTypesToChildren(types, allTypesHaveSameIdentity, children);
+
+            return children.ToImmutableAndFree();
+
+            static void addNamespacesToChildren(ArrayBuilder<SingleNamespaceDeclaration> namespaces, bool allNamespacesHaveSameName, ArrayBuilder<MergedNamespaceOrTypeDeclaration> children)
             {
-                if (allNamespacesHaveSameName)
-                {
-                    children.Add(MergedNamespaceDeclaration.Create(namespaces.ToImmutableAndFree()));
-                }
-                else
+                if (namespaces != null)
                 {
-                    var namespaceGroups = namespaces.ToDictionary(n => n.Name, StringOrdinalComparer.Instance);
-                    namespaces.Free();
-
-                    foreach (var namespaceGroup in namespaceGroups.Values)
+                    if (allNamespacesHaveSameName)
+                    {
+                        children.Add(MergedNamespaceDeclaration.Create(namespaces.ToImmutableAndFree()));
+                    }
+                    else
                     {
-                        children.Add(MergedNamespaceDeclaration.Create(namespaceGroup));
+                        // PERF: Don't use ArrayBuilder.ToDictionary directly as it requires an extra dictionary allocation. Other options such
+                        // as MultiDictionary<string, SingleNamespaceDeclaration> and Dictionary<string, OneOrMany<SingleNamespaceDeclaration>>
+                        // are even less appealing as they don't perform well when their value sets grow to contain a large number of items,
+                        // as typically happens when processing the namespaces.
```

Comment:
Perhaps we could add a `ToDictionary` overload that would do this? Or even make the current one to not allocate twice, I'm not sure what's the advantage of that.

Reply:
The two allocs are needed if the caller wants a dictionary of immutablearrays. We don't need that, so I could add an extension that returns a dictionary of arraybuilders, but I don't think we'd want to add that directly to arraybuilder, but rather in an extension method, so it would unfortunately be a near duplicate of code living in the arraybuilder class. 

Anyways, I'm not certain that's better, so I'll go ahead and use what's in the PR. If I notice this usage pattern again, it might warrant the additional ToDictionary overload.